### PR TITLE
fix(kad): preserve deadline in keep alive logic

### PR DESCRIPTION
**Pull request by nathanielc**
## Description

Previous to this change if the ConnectionHandler::poll for kad was called more frequently than the connection idle timeout the timeout would continually be pushed further into the future. After this change kad now will preserve the existing idle deadline.

## Notes & open questions

How should I add tests?

## Change checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: protocols/kad/src/handler_priv.rs [hunk 1]
```diff
@@ -759,7 +759,13 @@ where
...
         if self.outbound_substreams.is_empty() && self.inbound_substreams.is_empty() {
             // We destroyed all substreams in this function.
-            self.keep_alive = KeepAlive::Until(Instant::now() + self.config.idle_timeout);
+            self.keep_alive = if let KeepAlive::Until(time) = self.keep_alive {
+                // Preserve the existing idle timeout
+                KeepAlive::Until(time)
+            } else {
+                // Setup idle timeout
+                KeepAlive::Until(Instant::now() + self.config.idle_timeout)
+            };
```
Comment (mxinden) [inline protocols/kad/src/handler_priv.rs]: Functionality wise, this looks good to me. What do you think of decreasing the nesting?

``` diff
-        if self.outbound_substreams.is_empty() && self.inbound_substreams.is_empty() {
-            // We destroyed all substreams in this function.
-            self.keep_alive = KeepAlive::Until(Instant::now() + self.config.idle_timeout);
-        } else {
-            self.keep_alive = KeepAlive::Yes;
-        }
+        let no_streams = self.outbound_substreams.is_empty() && self.inbound_substreams.is_empty();
+        self.keep_alive = match (no_streams, self.keep_alive) {
+            // No open streams. Preserve the existing idle timeout.
+            (true, k @ KeepAlive::Until(_)) => k,
+            // No open streams. Set idle timeout.
+            (true, _) => KeepAlive::Until(Instant::now() + self.config.idle_timeout),
+            // Keep alive for open streams.
+            (false, _) => KeepAlive::Yes,
+        };
```

**Review by mxinden — COMMENTED** [review]
Thanks for the patch!

**Comment by nathanielc** [comment]
@mxinden Thanks for the review. I agree I like the simpler logic. PR updated and rebased off master.

**Comment by nathanielc** [comment]
I looked into the CI failures and they seem to be unrelated to the change. One is a clippy error for code I did not change. The other is a failed test in gossipsub.

**Review by mxinden — APPROVED** [review]
Thanks for the help @nathanielc! :pray:

**Pull request merged**